src/framework/numerical_testing_type.F90

```diff
@@ -0,0 +1,371 @@
+  logical,  optional, intent(in) :: ignore !< If present and true, ignore a fail
+  ! Local variables
+  integer :: k
+  logical :: this_test, ignore_this_fail
+  real :: tolerance, err ! Tolerance and error [A]
+
+  tolerance = 0.0
+  if (present(tol)) tolerance = tol
+  ignore_this_fail = this%ignore_fail
+  if (present(ignore)) ignore_this_fail = ignore
+  this_test = .false.
+
+  ! Scan for any mismatch between u_test and u_true
+  do k = 1, n
+    if (present(robits)) tolerance = abs(u_true(k)) * float(robits) * epsilon(err)
+    if (abs(u_test(k) - u_true(k)) > tolerance) this_test = .true.
+  enddo
+
+  ! If either being verbose, or an error was measured then display results
+  if (this_test .or. this%verbose) then
+    write(this%stdout,'(a4,2a24,1x,a)') 'k','Calculated value','Correct value',label
+    if (this_test) write(this%stderr,'(a4,2a24,1x,a)') 'k','Calculated value','Correct value',label
+    do k = 1, n
+      if (present(robits)) tolerance = abs(u_true(k)) * float(robits) * epsilon(err)
+      err = u_test(k) - u_true(k)
+      if ( ( abs(err) > tolerance .and. ignore_this_fail ) .or. &
+           ( abs(err) > 0. .and. abs(err) <= tolerance ) ) then
+        write(this%stdout,'(i4,1p2e24.16,a,1pe24.16,a)') k, u_test(k), u_true(k), &
+                         ' err=', err, ' <--- IGNORING'
+      elseif (abs(err) > tolerance) then
+        write(this%stdout,'(i4,1p2e24.16,a,1pe24.16,a)') k, u_test(k), u_true(k), &
+                         ' err=', err, ' <--- WRONG'
+        write(this%stderr,'(i4,1p2e24.16,a,1pe24.16,a)') k, u_test(k), u_true(k), &
+                         ' err=', err, ' <--- WRONG'
+      else
+        write(this%stdout,'(i4,1p2e24.16)') k, u_test(k), u_true(k)
+      endif
+    enddo
+  endif
+
+  call this%test( this_test, label, ignore=ignore_this_fail ) ! Updates state and counters in this
+end subroutine real_arr
+
+!> Compare i_test to i_true and report and return true if a difference is found
+!!
+!! If in verbose mode, display results to stdout
+!! If a difference is measured, display results to stdout and stderr
+subroutine int_arr(this, n, i_test, i_true, label, ignore)
+  class(testing),     intent(inout) :: this   !< This testing class
+  integer,               intent(in) :: n      !< Number of cells in u
+  integer, dimension(n), intent(in) :: i_test !< Values to test [A]
+  integer, dimension(n), intent(in) :: i_true !< Values to test against (correct answer) [A]
+  character(len=*),      intent(in) :: label  !< Message
+  logical,  optional,    intent(in) :: ignore !< If present and true, ignore a fail
+  ! Local variables
+  integer :: k
+  logical :: this_test, ignore_this_fail
+
+  ignore_this_fail = this%ignore_fail
+  if (present(ignore)) ignore_this_fail = ignore
+  this_test = .false.
+
+  ! Scan for any mismatch between u_test and u_true
+  do k = 1, n
+    if (i_test(k) .ne. i_true(k)) this_test = .true.
```

Comment:
Elsewhere in the MOM6 code, we use `/=` instead of `.ne.` for inequality tests.

Reply:
Fixed, thx